### ISSUE TITLE
Making custom CUDA function for R torch

### ISSUE THREAD

**Issue by dominikj2** [mlverse/torch#608]
Hi,

I am translating a pytorch code into R torch but it uses Pytorch functions with CUDA extensions. See cuda_op file in:

[https://github.com/lilanxiao/Rotated_IoU]()

In summary, the code is a loss function (Differentiable) with varying IoU (DIoU, GIoU etc) for rotating (and axis aligned) 3D Boxes. 

I am not sure how to use R torch code with CUDA extensions? 
What do I change in the cuda op file, which consists of CUDA and (C++ or C) wrappers ? 
I am not able to find an examples that demonstrate implementation of a custom CUDA function with R torch (see [ https://github.com/chrischoy/pytorch-custom-cuda-tutorial](url) and [https://discuss.pytorch.org/t/are-there-any-tutorial-for-customizing-functional-using-c-with-cuda/7522](url) for pytorch examples). If one doesn't exist, can anyway demonstrate such an example using the Rotated_IoU code?

Thank you.

**Comment by dfalbel** [comment]
Hi @dominikj2,

We currently don't support extending torch with c++/cuda operations from external packages. We plan to work on this soon. In the meantime, one thigh we could do is adding a `contrib` function to torch. 

See for example this PR: https://github.com/mlverse/torch/pull/309/files that adds contrib_sparsemax to torch.

Let me know if you want to try a PR, otherwise I can add the Rotated_IoU in the next few days.

**Comment by dominikj2** [comment]
Hi @dfalbel,

That sounds great. I would like to try a PR and start contributing more.

I am still in the process of translating the python Rotated_IoU into torch for R. I will be in touch when completed. Thank you for all your efforts!

**Comment by dominikj2** [comment]
Hi @dfalbel,

I have finished translating Rotated_IoU Loss from the following python code: [https://github.com/lilanxiao/Rotated_IoU]()

Below is the attached code as a txt file.
[TORCH_ROTATING_3D_MAIN.txt](https://github.com/mlverse/torch/files/6829942/TORCH_ROTATING_3D_MAIN.txt)

It crashes when it needs to use the Cuda_op function `sort_v` (i.e. c++/cuda operations).

Helping me add a `contrib` function to torch that integrates the Cuda_op function would be appreciated. Once the code runs past `sort_v` error, I will need to debug thereafter.

I would like to get involved anyway possible so that I could perform such tasks in the future. Please let me know what I can do to develop such skills. 

Thank you!

**Comment by dominikj2** [comment]
Hi @dfalbel,
To address the fact that Torch does not support extending c++/cuda operations from external packages, I would like to follow up your proposition to "adding a contrib function to torch"? 

 I have mentioned that I am interested in trying a PR (this will be my first) but have not received a response. I am not sure if I am expected to navigate this myself without any guidance or whether you proposed to set up a PR for this purpose?

I am not sure what you mean by " I can add the Rotated_IoU function to torch"?  I may have wrongly assumed you would use the code I provided in the txt file and integrate the  c++/cuda operations in the PR. Can you clarify?

I have been eagerly learning c++/cuda operations and wonder when you anticipate Torch will support "extending c++/cuda operations from external packages"? 

Thanks

**Comment by dfalbel** [comment]
Hi @dominikj2,

Sorry for the late reply, please see the #619 PR. 

I copied from the `cuda_op` directory in the repository you linked and added cmake tooling in lantern to make sure this get correctly compiled. 

I listed a few work items to finalize that PR.

In order to make use of it you will need to compile lantern from source. You should be able to do it with:

- Have a CMake installation
- From the torch directory run: `source("tools/buildlantern.R")`
- Run `devtools::load_all()`

Let me know if I can help with something! Also happy to pair on the Discord channel so we can talk and share screens to do it together.

**Comment by dominikj2** [comment]
Hi @dfalbel,

I've installed CMake following this 2 min video [https://www.youtube.com/watch?v=8_X5Iq9niDE](url). In summary, (1) downloaded Windows x64 ZIP | cmake-3.21.0-windows-x86_64.zip, (2) unzipped, and (3) put directory in systems environment variables.

Regarding second step, my torch directory (C:\RFolder\R\win-library\4.0\torch) does not have a tools folder so the source command doesn't work. In R, I typed source("C:/RFolder/R/win-library/4.0/torch/tools/buildlantern.R") and R obviously couldn't find file. I'm probably misunderstanding second instruction. Could you clarify or provide suggestions?
Thanks

**Comment by dominikj2** [comment]
Hi @dfalbel,

I assumed above error may have been due to you making changes to `devtools::install_github("dfalbel/torch")` so I tried to re-install it and it gave me the below error. Is my `tools/buildlantern.R` error related to below installation error. I have previous torch version that I will revert back to. Do I need to reinstall newer `dfalbel/torch` version to access the `tools/buildlantern`?

Also, Iet me know if you would prefer to solve this problem on the Discord channel. Let me know your most convenient time and I will aim to meet.

```
torch_types.h:3:10: fatal error: torch/torch.h: No such file or directory
 #include <torch/torch.h>
          ^~~~~~~~~~~~~~~
compilation terminated.
make: *** [C:/PROGRA~1/R/R-40~1.2/etc/i386/Makeconf:229: RcppExports.o] Error 1
ERROR: compilation failed for package 'torch'
* removing 'C:/RFolder/R/win-library/4.0/torch'
Warning message:
In i.p(...) :
  installation of package ‘C:/Users/dj806/AppData/Local/Temp/1/RtmpqCy4rW/file36882d51758b/torch_0.0.1.9000.tar.gz’ had non-zero exit status
```

**Comment by dfalbel** [comment]
Hi @dominikj2,

Sorry I wasn't clear, in order to further develop it, you need to:

- Clone this repository to your computer, you can use `usethis::create_from_github("mlverse/torch")` to do this.
- Checkout into the PR branch. You can do it using: `usethis::pr_fetch(number = 619)`

Now you should have an RStudio window opened in the torch project.
You can now run `source("tools/buildlantern.R")` and `devtools::load_all())`.
